Visual improvements in sublayers dialog

suricactus opened the pull request:
An annoying lack of context when adding multiple layers:
1) select multiple multi-layer geopackage layers from a directory in the file browser
2) drag and drop them in QGIS
3) the "Select Layers to Add..." dialog does not give any hint about the source file

This PR fixes 3) by adding the current file name in the window title.
![Peek 2020-03-14 01-11](https://user-images.githubusercontent.com/2820439/76666233-5e00de80-6591-11ea-9416-ba2fda0bb353.gif)

Fixes: https://github.com/qgis/QGIS/issues/26905

suricactus committed:
Add the filename to the "Select Layers to Add.."

elpaso commented:
Nice! 
What do you think about hiding the path and showing the filename only? 
I'm worried about long paths.

suricactus commented:
Imagine a situation you import files from a file tree, where you have multiple files with the same name. 
```
./census/2001/filename.gpkg
./census/2011/filename.gpkg
./census/2021/filename.gpkg
```
We should be covered  for this situation, I don't think it's that uncommon.

A possible solution would be to use, so we can shorten the path in most cases:
```
QDir( QgsProject::instance().homePath() ).relativeFilePath("D:/images/image.bmp");
```
Unfortunately storing a list of the current filenames to be shown is not that easy and trivial. I will add something like this is in `handleDropUriList`, where the suggested filename is generated.

Any other solutions is appreciated.

suricactus committed:
Sublayer selection dialog improvements
- Disable ok button if there are no layers selected
- Select all layers by default 
- Remove filename prefix in the layer tree
- Show the imported filename in the dialog title bar

suricactus committed:
Make sublayer prompt behavior the same for all types of layers

suricactus commented:
A second though of the use case described in my post above. In such situation the files are most likely to have the same structure and one would like to import the same layers. This is just an excuse, the truth is I don't find an easy way to fix. The idea to use `relativeFilePath` will produce ugly "../../../../" strings and you are right that `/home/suricactus/projects/very long project name/layer/census/2010 unofficial/demographics/by age and gender/united states of america.gpkg` will look terrible in the title bar. 

I am using this branch to apply other small fixes. Summary:
- Fixes https://github.com/qgis/QGIS/issues/26905 **Show currently added filename in the titlebar**.
- Fixes https://github.com/qgis/QGIS/issues/26670 **Remove the filename prefix in the "Layer Tree"**. Useless visual clutter. Putting all the layer in a group is already present, so there is no risk of mixing layer names together. Adding this as an additional checkbox in the dialog causes will make the interface to look complicated.
- **Disable the OK button if there are no layers selected.** Most users would just click "OK" and would be confused why there is no effect.
- **Select all layers by default.** To finish the operation is exactly one mouse click of the "OK" button, as it is a common case to "Select All". Otherwise the user is always obliged to click on a layer or "Select All" and then to click "OK". 

Breaking changes:
- setting `qgis/promptForRasterSublayers` is renamed to `qgis/promptForSublayers`. The default value is still "Always". I renamed the setting to something sensible, there will be minor regression for some users who changed this value. The setting is renamed to "Prompt for ~raster~ sublayers when opening". It seems that this option either never existed in the UI, or has disappeared somehow.

Visual changelog:
- Settings:

![image](https://user-images.githubusercontent.com/2820439/76891680-0fee2280-6892-11ea-9406-3a55bb996361.png)

- In action:
![Peek 2020-03-17 23-14](https://user-images.githubusercontent.com/2820439/76902710-42554b00-68a5-11ea-9c8b-ea74d709b0c2.gif)

Known issues:
- **Add option to "Add to a group" zipfiles** - this option is a bit tricky to add, I have overload several functions. I would prefer to do that in a separate PR.

uclaros commented:
Regarding the long paths, a read only widget could be added above the table view (a grayed out lineedit?) like and address bar so you can also select and copy part of the path-filename.

suricactus commented:
I don't see the point of being able to copy your file path in 99% of the cases? You only need indication what are you currently seeing, no need to make the interface heavier than needed with add additional boxes in my opinion.

uclaros commented:
- It's not about copying it, it's about seeing it as you have already mentioned. Imho, having this info only on the titlebar is not a good choice, since the titlebar is customizable by the window manager it is not guaranteed that what is there is readable. Adding a single line of text is not going to make this simple dialog bloated. And in general it is good practice to be able to copy displayed paths. If it's there, why not allow copying it for the 1% case?

- If `PromptNever` is chosen it silently skips loading a file? Do we need that?

suricactus commented:
1) fair point on window managers. I will change that.
2) this was already existing behavior. I see it as a way to disable drag and drop of layer files in QGIS.

suricactus committed:
Add full path yet again. Remove dynamic form elements creation.

suricactus commented:
Looks a bit ugly, but now the user knows exactly what is being imported. 
![Peek 2020-03-19 06-31](https://user-images.githubusercontent.com/2820439/77031812-5f744180-69ab-11ea-8d9b-ee026ffcf89d.gif)

uclaros commented:
Removing the 'Path' QLabel and letting the QLineEdit occupy the full width will definitely make it look better.
You could also leave the filename in the titlebar too.

suricactus commented:
This will need changes here:
https://docs.qgis.org/testing/en/docs/user_manual/introduction/general_tools.html#identify

uclaros commented:
> This will need changes here:
> https://docs.qgis.org/testing/en/docs/user_manual/introduction/general_tools.html#identify

I guess this belongs to PR #35190

suricactus commented:
It took me some time to find the documentation error, the message is a bit hidden, even though colored in yellow.

@uclaros thanks for correction, something happened and I had a lot of comments posted in the wrong threads.

As for this PR:
- I agree it looks better without the "Path" label, so I removed it
- Put filename only in the window title

Yet another iteration of this gif:
![Peek 2020-03-25 07-36](https://user-images.githubusercontent.com/2820439/77505262-5b449a00-6e6b-11ea-9d79-ab720e473571.gif)

suricactus committed:
Remove the "Path" label; Add the filename in dialog's title

uclaros commented:
It looks very good! Is text selectable in the qlineedit?

suricactus committed:
File path selectable; US English typo

suricactus commented:
Now it is. :)

m-kuhn reviewed (commented):
If others have already left their :heart: for the UX it provides. Here are some code (style) level things to check and then this should be good to roll :+1:

suricactus commented on src/gui/qgssublayersdialog.cpp:
```diff
@@ -61,27 +63,34 @@ QgsSublayersDialog::QgsSublayersDialog( ProviderType providerType, const QString
 
+  QString fileFullPath = providerType == QgsSublayersDialog::Vsifile
+                         ? providerSource
+                         : QgsProviderRegistry::instance()->decodeUri( name, providerSource )[QStringLiteral( "path" )].toString();
```
done!

suricactus commented on src/gui/qgssublayersdialog.h:
```diff
@@ -93,7 +122,7 @@ class GUI_EXPORT QgsSublayersDialog : public QDialog, private Ui::QgsSublayersDi
      * \since QGIS 3.0
      */
-    bool addToGroupCheckbox() const { return mCheckboxAddToGroup->isChecked(); }
+    bool addToGroupCheckbox() const { return cbxAddToGroup->isChecked(); }
```
This is quite inconsistent across the codebase though. Renamed it to `mCbxAddToGroup`

m-kuhn replied:
"organically grown" is how you could call it...

suricactus replied:
How far should I go for this? Is this convention valid only for elements that are used in the code? There are quite a lot of GUI elements, that we never interact in "hand-written" code like layouts, separators, dialog buttons etc. Should we also change them?

I believe this organically grown change would be more widely applied if it is described somewhere. If you can point me documentation/style guide/github page/chat where this has been decided and described, would be nice. I would love to write these two sentences describing the requirement, but it is still unclear how it should be applied. Maybe here is a good candidate to put such recommendation? https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#qt-designer

m-kuhn replied:
> Is this convention valid only for elements that are used in the code? There are quite a lot of GUI elements, that we never manually use in the code like layouts, separators, dialog buttons etc. Should we also change them?

Yes, only for the ones used in code

> Maybe here is a good candidate to put such recommendation? https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#qt-designer

That's exactly where it should go, would be really nice to add that there!

suricactus commented:
@m-kuhn is it possible to reopen the PR or I have to make a new one?
